File: examples/go/core/ui.case.go
package core

import "app-protocol/examples/go/core/shared"

type UiContext struct {
	shared.AppBaseContext
	Renderer any
	Router   any
	Store    any
	API      shared.AppHttpClient
	Packages shared.Dict
	Extra    shared.Dict
}

type UiCaseSupport[TState any] struct {
	Ctx   *UiContext
	State TState
}

func NewUiCaseSupport[TState any](ctx *UiContext, initialState TState) UiCaseSupport[TState] {
	return UiCaseSupport[TState]{
		Ctx:   ctx,
		State: initialState,
	}
}

func (u *UiCaseSupport[TState]) SetState(next TState) {
	u.State = next
}

func (u *UiCaseSupport[TState]) Test() error {
	return nil
}

File: examples/kotlin/core/ui.case.kt
package app.protocol.examples.kotlin.core.ui

import app.protocol.examples.kotlin.core.Dict
import app.protocol.examples.kotlin.core.shared.AppBaseContext
import app.protocol.examples.kotlin.core.shared.AppHttpClient

interface UiContext : AppBaseContext {
    val renderer: Any?
    val router: Any?
    val store: Any?
    val api: AppHttpClient?
    val packages: Dict<Any?>?
    val extra: Dict<Any?>?
}

typealias UIState = MutableMap<String, Any?>

abstract class BaseUiCase<TState : Any>(
    protected val ctx: UiContext,
    initialState: TState,
) {
    protected var currentState: TState = initialState
        private set

    protected val state: TState
        get() = currentState

    abstract fun view(): Any

    open suspend fun test() {
    }

    protected open fun _viewmodel(vararg args: Any?): Any? = null

    protected open suspend fun _service(vararg args: Any?): Any? = null

    protected open suspend fun _repository(vararg args: Any?): Any? = null

    protected fun setState(nextState: TState) {
        currentState = nextState
    }
}
